tests/ui/needless_pass_by_value.rs

```diff
@@ -79,4 +79,27 @@ fn issue_2114(s: String, t: String, u: Vec<i32>, v: Vec<i32>) {
     let _ = v.clone();
 }
 
+struct S<T, U>(T, U);
+
+impl<T: Serialize, U> S<T, U> {
+    fn foo(
+        self, // taking `self` by value is always allowed
+        s: String,
+        t: String,
+    ) -> usize {
+        s.len() + t.capacity()
+    }
+
+    fn bar(
+        _t: T, // Ok, since `&T: Serialize` too
+    ) {
+    }
+
+    fn baz(
+        &self,
+        _u: U,
+    ) {
+    }
```

Comment:
can you add a test for another argument that is of type `Self`? Just that we document this behaviour?